File: leetcode/ValidParentheses/isValid.java
public class Solution {
    public boolean isValid(String s) {
        // Start typing your Java solution below
        // DO NOT write main() function
        Stack<Character> st = new Stack<Character>();
        HashMap<Character, Character> map = new HashMap<Character, Character>();
        map.put('(', ')');
        map.put('[', ']');
        map.put('{', '}');
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (c == '(' || c == '[' || c == '{') st.push(c);
            else if (st.empty() || map.get(st.pop()) != c) return false;
        }
        return st.empty();
    }
}

File: leetcode/ValidParentheses/isValid.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> st;
        for (char c : s) {
            if (c == '(' || c == '[' || c == '{') st.push(c);
            else {
                if (st.empty()) return false;
                char top = st.top();
                switch(c) {
                    case ')': if (top != '(') return false; break;
                    case ']': if (top != '[') return false; break;
                    case '}': if (top != '{') return false; break;
                    default: return false;
                }
                st.pop();
            }
        }
        return st.empty();
    }
};
